nussl/datasets/hooks.py

```diff
@@ -417,6 +421,168 @@ def process_item(self, item):
+                 threshold_db: float = -60.0,
+                 segment_dur: float = 4.0,
+                 hop_ratio: float = 0.5,
+                 verbose: bool = False,
+                 padding_mode: int = 0,
+                 **kwargs):
+        assert sample_rate is not None, f"Please provide a sample rate for the dataset."
+        self.threshold_db = threshold_db
+        self.segment_dur = segment_dur
+        self.hop_ratio = hop_ratio
+        self.sample_rate = sample_rate
+        self.salient_src = salient_src
+        self.folder = folder
+        self.verbose = verbose
+        self.padding_mode = padding_mode
+
+        self.mix_src = MixSourceFolder(
+            folder, mix_folder, source_folders, ext, make_mix, **kwargs
+        )
+        self.song_metadata = self._populate_metadata()
+        self.offset = 0.0
+
+
+        super(SalientExcerptMixSourceFolder, self).__init__(
+            self._get_mix,
+            len(self.song_metadata),
+            sample_rate=sample_rate,
+            **kwargs
+        )
+
+    def _populate_metadata(self):
+        """Makes a metadata dict of salient start times for each source."""
+
+        if self.verbose:
+            # TODO: logging.info()
+            print(f'Preprocessing data at {self.folder}.')
+
+        # TODO: Balanced dataset!
+        items = self.mix_src.get_items(self.folder)
+        metadata = []
+        for item in tqdm.tqdm(items, disable=not self.verbose):
+            mix, sources = self.mix_src.get_mix_and_sources(item)
+            target_src = sources[self.salient_src]
+            # Todo: Repair omission code
+            # if len(target_src) / self.sample_rate < self.segment_dur:
+            #     if self.padding_mode == 1:
+            #         # omit the current item
+            #         continue
+            target_audio = target_src.audio_data
+            salient_starts = utils.find_salient_starts(target_audio,
+                                                       self.segment_dur,
+                                                       self.hop_ratio,
+                                                       self.sample_rate,
+                                                       self.threshold_db)
+            for start in salient_starts:
+                metadata.append({
+                    'mixsrc_item': item,
+                    'start': start
+                })
+        return metadata
+
+    def _get_mix(self, placeholder, item):
+        """OnTheFly closure. Gets called for every iteration to build a batch. A placeholder param is necessary since
+        OnTheFly will pass itself as well as the item to this function"""
+        item = self.song_metadata[item]
+        mixsrc_item = item['mixsrc_item']
+        # Note that the offset here is really the onset, but the parameter is passed as a kwarg to what is ultimately
+        # the AudioSignal which takes the onset as the parameter "offset"
+        offset = item['start']
```

Comment:
Did you test that this works with multiple threads? I.e., with a pytorch data loader with `n_workers > 1`? This seems like it could be a race condition, but I'm not 100% sure. If it is a race condition, then we need to rearchitect this object

Reply:
I have not tested it with multiple threads yet, I'll be sure to do that. It would have been a race condition given the initial implementation where offset was a class attribute. Changing it to a function variable should limit any possible collisions to be on read which is perfectly fine.

Reply:
Oh, yeah. Nice. I see what you did here. Yeah, that's a nice work around.